### ISSUE TITLE
Generate assets at the image build time

### ISSUE THREAD

**Issue by bqbn** [mozilla/addons-frontend#12031]
### Describe the problem and steps to reproduce it:

Currently, we run this command to start the frontend,

```
/bin/sh -c "yarn build && API_HOST=https://internal-api.addons.mozilla.org yarn start"
```

I believe `yarn build` is used to generate the static asset files. However, this step takes 2-5 minutes to finish depending on the size of the ec2 instance. This delays the start of the application quite a bit and it happens to all app instances.

The purpose of this issue is to see if we can run `yarn build` at the image build time so that it takes less time to start the application.

### What happened?
n/a

### Anything else we should know?

The following are the env variables that are passed to the container when

```
/bin/sh -c "yarn build && API_HOST=https://internal-api.addons.mozilla.org yarn start"
```

runs.

```
NODE_ENV=production
NODE_CONFIG_ENV=prod
SERVER_HOST=0.0.0.0
SERVER_PORT=4000
STATSD_HOST=localhost
STATSD_PORT=8125
```

**Comment by bqbn** [comment]
I'm not sure how `addons-code-manager` does it, but apparently that application is able to build the assets at the image build time, and it only runs `yarn start` (see https://github.com/mozilla/addons-code-manager/blob/master/Dockerfile#L42)  to start the app.

So maybe that app can be used as an example for this one.

**Comment by bobsilverberg** [comment]
Is the general idea here that we would run `yarn build` when creating the docker image, and therefore the files would already be in the image and we wouldn't have to run `yarn build` when starting addons-frontend?

**Comment by diox** [comment]
Yes - but AFAIK the build is specific to the env at the moment, so we would either need to pre-build all envs or find a way to ensure the exact same build is used everywhere.

**Comment by bobsilverberg** [comment]
A couple of follow up questions:
1. Aren't we already building the assets into the image [here](https://github.com/mozilla/addons-frontend/blob/master/.circleci/config.yml#L184) and [here](https://github.com/mozilla/addons-frontend/blob/master/.circleci/config.yml#L250)? Are we then overwriting them at startup in order to change them to match the environment?
2. Would having a separate image per environment be an option to address the fact that the builds need to be different?

**Comment by bqbn** [comment]
From ops perspective, we'd prefer one generic image and use a single environment variable to control what assets to use for each environment at container startup time.

**Comment by bobsilverberg** [comment]
@bqbn What are the exact commands and values for environment variables when we start the service on both stage and dev?

**Comment by bobsilverberg** [comment]
I have done a bit of experimenting with this. I am able to generate a different set of assets and store them in different folders, based on the value of `NODE_CONFIG_ENV`. I am trying to also get that to work for `webpack-assets.json` because it isn't written to the `dist` folder. I have been partially successful with that, but before I spend more time on it, does this seem like a good route to go? I.e., we would run a `build` command for all three environments that would generate all three sets of assets and configurations, and then store those all in the docker image?

**Comment by bqbn** [comment]
> @bqbn What are the exact commands and values for environment variables when we start the service on both stage and dev?

For dev, the command is

```
         /bin/sh -c "yarn build && API_HOST=https://internal-api.addons-dev.allizom.org yarn start"
```
 and the environment variables are,
```
NODE_ENV=production
NODE_CONFIG_ENV=dev
SERVER_HOST=0.0.0.0
SERVER_PORT=4000
STATSD_HOST=localhost
STATSD_PORT=8125
PORT=4000
```

For stage, the command is,
```
         /bin/sh -c "yarn build && API_HOST=https://internal-api.addons.allizom.org yarn start"
```
and the environment variables are,
```
NODE_ENV=production
NODE_CONFIG_ENV=stage
SERVER_HOST=0.0.0.0
SERVER_PORT=4000
STATSD_HOST=localhost
STATSD_PORT=8125
PORT=4000
```

**Comment by bobsilverberg** [comment]
I did a bit of digging into how we might be able to have a single bundle that works for all environments, as opposed to needing different bundles for each environment. From what I can tell, the only difference between the bundles are a few variables that get substituted in the `config`, e.g., `apiHost` is `https://addons.mozilla.org` for prod, and `https://addons.allizom.org` for stage.

I started to experiment with having all the values for each environment stored in a single config, and then dynamically finding the right one based on the current environment, but then I discovered that we can override the value of most of these via environment variables with we run `yarn start`. So, for example, when starting on prod we could include `API_HOST=https://addons.mozilla.org` and when starting on stage we could include `API_HOST=https://addons.allizom.org`. One downside to this is that these values are no longer stored in the config but rather need to be specified when issuing the commands to start the instances. The upside is that this seems like a very simple way of addressing the issue.

@diox @eviljeff @bqbn What do you think?

**Comment by eviljeff** [comment]
> What do you think?

seems good to me - I guess we would (could) have defaults when running these commands locally?  (or separate scripts for local dev, I suppose)

**Comment by KevinMind** [comment]
Old Jira Ticket: https://mozilla-hub.atlassian.net/browse/ADDFRNT-206
